internals/controllers/subnamespace/subnamespace_sync.go

```diff
@@ -0,0 +1,407 @@
+			ResourceQuotaSpec: sns.Spec.ResourceQuotaSpec,
+		}
+		childrenRequests = append(childrenRequests, x)
+
+		snsRequest := sns.Spec.ResourceQuotaSpec.Hard
+		for res := range sns.Spec.ResourceQuotaSpec.Hard {
+			var (
+				totalRequest, _ = allocated[res]
+				vRequest, _     = snsRequest[res]
+			)
+			allocated[res] = *resource.NewQuantity(vRequest.Value()+totalRequest.Value(), resource.BinarySI)
+		}
+	}
+
+	return childrenRequests, allocated
+}
+
+// computeSNSFreeToAllocate computes the resources that are still free to allocate by
+// looking at the total available resources in the subnamespace spec and the currently allocated resources
+func computeSNSFreeToAllocate(snsObject *utils.ObjectContext, allocated corev1.ResourceList) corev1.ResourceList {
+	var freeToAllocate = corev1.ResourceList{}
+
+	for res := range snsObject.Object.(*danav1.Subnamespace).Spec.ResourceQuotaSpec.Hard {
+
+		var (
+			totalRequest, _ = allocated[res]
+			vRequest, _     = snsObject.Object.(*danav1.Subnamespace).Spec.ResourceQuotaSpec.Hard[res]
+		)
+		value := vRequest.Value() - totalRequest.Value()
+		freeToAllocate[res] = *resource.NewQuantity(value, resource.BinarySI)
+	}
+
+	return freeToAllocate
+}
+
+// deleteLegacyCRQ deletes a pre-exising CRQ in cases where they should only be a RQ.
+// This is needed because historically we used to only have CRQs and didn't introduce RQs until later
+func deleteLegacyCRQ(snsObject *utils.ObjectContext) error {
+	if utils.DoesSNSCrqExists(snsObject) {
+		crqName := snsObject.Object.GetName()
+		crqObj, err := utils.NewObjectContext(snsObject.Ctx, snsObject.Client, types.NamespacedName{Name: crqName}, &quotav1.ClusterResourceQuota{})
+		if err != nil {
+			return err
+		}
+		if err := crqObj.EnsureDeleteObject(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// syncSNSAnnotations sync subnamespace annotations
+func syncSNSAnnotations(snsObject, snsParentNS, parentSNS *utils.ObjectContext, isRq bool) error {
+	annots := snsObject.Object.GetAnnotations()
+
+	if isRq {
+		annots[danav1.IsRq] = danav1.True
+	} else {
+		annots[danav1.IsRq] = danav1.False
+	}
+
+	if upperRPName, isUpper, err := GetUpperResourcePoolName(snsObject, parentSNS); err != nil {
+		return err
+	} else {
+		annots[danav1.UpperRp] = upperRPName
+		annots[danav1.IsUpperRp] = isUpper
+	}
+
+	annots[danav1.CrqPointer] = utils.GetCrqPointer(snsObject.Object)
+
+	displayName := utils.GetNamespaceDisplayName(snsParentNS.Object) + "/" + snsObject.Object.GetName()
+	annots[danav1.OpenShiftDisplayName] = displayName
+	annots[danav1.DisplayName] = displayName
+
+	if err := snsObject.AppendAnnotations(annots); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetUpperResourcePoolName returns the name of the upper ResourcePool of a given subnamespace
+// and returns true if the SNS is the upper-rp and false otherwise
+func GetUpperResourcePoolName(sns *utils.ObjectContext, parentSNS *utils.ObjectContext) (string, string, error) {
```

Comment:
The is upper should return here as bool for my case only in the api level it should be string